### PR TITLE
Custom implementation for node-fetch, handling cookies during redirections

### PR DESCRIPTION
**Pull request by mvindahl**

### CODE REVIEW

**Comment by mvindahl** [comment]
I took a shot at documenting the functionality in the README. I will not take offense if you decide to rephrase or change stuff. I believe that you have been granted edit rights to my master branch.

**Comment by valeriangalliat** [comment]
Thanks a lot, that looks good to me! I wish the redirect logic didn't have to be part of this library, like if `node-fetch` could have a `self` option taking a wrapper function to call for redirects, this way it would avoid duplicating the redirect logic and keep it and maintain it only in `node-fetch`.

But for now I think that solution is our best option, maybe we can aim at adding that option in `node-fetch` if you think it would make sense too?

Also thanks for updating the readme, it's perfect. I'll just make a couple style updates on the code after merging. :)

File: node-fetch.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+                // interpret the proprietary "redirect" option in the same way that node-fetch does
+                if (isRedirect && opts.redirect !== 'manual' && opts.follow !== 0) {
+                    var optsForGet = Object.assign({}, {
+                        method: 'GET',
+                        body: null,
+                        // since the "follow" flag is not relevant for node-fetch in this case,
+                        // we'll hijack it for our internal bookkeeping
+                        follow: opts.follow !== undefined ? opts.follow - 1 : undefined
+                    });
+                    return nodeFetchCookie(res.headers._headers.location[0], optsForGet);
```
Comment (valeriangalliat) [inline node-fetch.js]: Why use `headers._headers.location[0]` instead of `headers.get('location')`?

Comment (mvindahl) [inline node-fetch.js]: Mistake on my part, sorry. Didn't check the API when I wrote the initial implementation, just dumped the response object. Your suggestion is the correct one, of course.

**Comment by mvindahl** [comment]
I agree that this is a cumbersome way to accomplish stuff, and that replicating logic from node-fetch is a bit smelly. In fact, I kind of wish that node-fetch would incorporate cookie handling logic internally as that would probably make the whole thing more elegant and would remove the need for patching it with a fetch-cookie library.

There has been some discussion on approaches on the node-fetch bugtracker. As far as I can gather, they have made a design decision to not incorporate cookies, and they have also made a design decision to not implement any kind of callback/event mechanism :-/

Refs:
https://github.com/bitinn/node-fetch/issues/146
https://github.com/bitinn/node-fetch/issues/94

File: node-fetch.js (hunk 1 above)
Comment (valeriangalliat) [inline node-fetch.js]: No worries, just wanted to make sure :)

**Comment by valeriangalliat** [comment]
`fetch-cookie/node-fetch` is published as of 0.4.0 :)

Thanks for the refs. If there's no way cookie handling will be added to node-fetch, it would still be very easy for node-fetch for support wrappers by adding a `self` option and using it when recursively calling itself in the redirect logic. I'm going to make a PR for this.

**Comment by mvindahl** [comment]
I like your PR for node-fetch, let's see if they accept it. It would of course mean a rewrite of pretty much all the code that I submitted. But it would lead to a better solution :)

Cheering for you.

**Pull request merged**